File: backend/app/schemas/analytics.py
from datetime import datetime, date
from typing import List, Optional, Dict, Any
from uuid import UUID
from pydantic import BaseModel, ConfigDict

class DashboardWidgetResponse(BaseModel):
    id: UUID
    dashboard_id: UUID
    title: str
    widget_type: str
    size_x: int
    size_y: int
    position_x: int
    position_y: int
    query_config: Dict[str, Any]
    model_config = ConfigDict(from_attributes=True)

class DashboardResponse(BaseModel):
    id: UUID
    name: str
    code: str
    allowed_roles: str
    description: Optional[str]
    model_config = ConfigDict(from_attributes=True)

class KPIDefinitionResponse(BaseModel):
    id: UUID
    name: str
    metric_code: str
    target_value: float
    threshold_yellow: float
    threshold_red: float
    status: str
    model_config = ConfigDict(from_attributes=True)

class KPIValueResponse(BaseModel):
    id: UUID
    kpi_id: UUID
    current_value: float
    status_indicator: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class AnalyticsQueryRequest(BaseModel):
    metric_code: str
    dimensions: List[str] = []
    filters: Dict[str, Any] = {}

class AnalyticsQueryResponse(BaseModel):
    metric: str
    value: float
    dimensions: List[str]
    filters: Dict[str, Any]
    timestamp: str

class DeterministicInsightResponse(BaseModel):
    category: str
    severity: str
    message: str
    action_url: Optional[str] = None

class ForecastResultResponse(BaseModel):
    date: str
    value: float
    lower: float
    upper: float

File: frontend/src/types/analytics.ts
export interface BusinessMetric {
  id: string;
  name: string;
  code: string;
  formula: string;
  dependencies_json: Record<string, any>;
  cache_refresh_rate: 'DAILY' | 'HOURLY' | 'REALTIME';
  description?: string;
}

export interface KPIDefinition {
  id: string;
  name: string;
  metric_code: string;
  target_value: number;
  threshold_yellow: number;
  threshold_red: number;
  status: string;
}

export interface KPIValue {
  id: string;
  kpi_id: string;
  current_value: number;
  status_indicator: 'GREEN' | 'YELLOW' | 'RED';
  created_at: string;
}

export interface DeterministicInsight {
  category: string;
  severity: 'INFO' | 'WARNING' | 'CRITICAL';
  message: string;
  action_url?: string;
}

export interface ForecastResult {
  date: string;
  value: number;
  lower: number;
  upper: number;
}

export interface Dashboard {
  id: string;
  name: string;
  code: string;
  allowed_roles: string;
  description?: string;
}
